Controllers/controller_mpc_tf.py

```diff
@@ -0,0 +1,61 @@
+import os
+from importlib import import_module
+from typing import Optional
+from SI_Toolkit.Predictors.predictor_wrapper import PredictorWrapper
+
+import numpy as np
+import yaml
+from Control_Toolkit.Controllers import template_controller
+from Control_Toolkit_ASF.Cost_Functions import cost_function_base
+from gym.spaces.box import Box
+
+from Control_Toolkit.Optimizers import template_optimizer
+from SI_Toolkit.computation_library import (NumpyLibrary, PyTorchLibrary,
+                                                TensorFlowLibrary, TensorType)
+from Control_Toolkit.others.globals_and_utils import get_logger, import_optimizer_by_name
+
+
+config_optimizer = yaml.load(open(os.path.join("Control_Toolkit_ASF", "config_optimizers.yml")), Loader=yaml.FullLoader)
+config_controller = yaml.load(open(os.path.join("Control_Toolkit_ASF", "config_controllers.yml")), Loader=yaml.FullLoader)
+config_cost_function = yaml.load(open(os.path.join("Control_Toolkit_ASF", "config_cost_function.yml")), Loader=yaml.FullLoader)
+
+
+logger = get_logger(__name__)
+
+
+class controller_mpc_tf(template_controller):
+    _computation_library = TensorFlowLibrary
+    _has_optimizer = True
+    
+    def configure(self, optimizer_name: Optional[str]=None):
+        if optimizer_name in {None, ""}:
+            optimizer_name = str(config_controller["mpc-tf"]["optimizer"])
+            logger.info(f"Using default optimizer {optimizer_name} specified in config file")
+        
+        # Create cost function
+        cost_function_name: str = config_cost_function["cost_function_name"]
+        cost_function_module = import_module(f"Control_Toolkit_ASF.Cost_Functions.{self.environment_name}.{cost_function_name}")
+        self.cost_function: cost_function_base = getattr(cost_function_module, cost_function_name)(self, self.computation_library)
+        
+        # Create predictor
+        self.predictor = PredictorWrapper()
+        
+        # MPC Controller always has an optimizer
+        Optimizer = import_optimizer_by_name(optimizer_name)
+        self.optimizer: template_optimizer = Optimizer(
+            controller=self,
```

Comment:
The motivation was logging variables that the optimizer produces, like the input plans or cost values.

My implementation makes the optimizer the "publisher" of logging information and sends it to the controller "subscriber". However, it's good you're raising this because I personally didn't like passing controller to optimizer.

Let me suggest this: The controller is responsible for grabbing logging values from its optimizer and assumes they exist if needed. If for some reason it can't find them in the optimizer instance, it doesn't break.

Reply:
Implemented in cd7f7d5

Reply:
Sounds great.